### PR TITLE
Fix resource failure, change requirements handling, and refactor test config

### PR DESCRIPTION
**Pull request by johnsca**
It turns out that even though Juju claims that it is attaching resources when you deploy a bundle referring to store charms w/ resources along with an overlay that points to local copies of the charms, it actually doesn't. So the resources were failing, but because of how the handlers were working, the status was not being surfaced properly and the charm seemed like it was functional when it wasn't.

This also switches to pipenv to manage and lock the requirements, with an update target for updating the dependencies. It also switches the operator framework to pull from GitHub rather than PyPI to pick up some newer, as yet unreleased fixes (specifically the config default value handling in the test harness). Finally, it refactors the tox config to use pipenv, remove the unused build env, and makes running tox in the repo root actually run the tests for the charms.

### CODE REVIEW

File: charms/metallb-speaker/src/charm.py [hunk 1]
```diff
@@ -22,97 +22,72 @@
...
-        )
-        utils.delete_namespaced_role_with_api(
-            name='pod-lister',
-            namespace=self._stored.namespace
-        )
-        self.model.unit.status = ActiveStatus("Removing extra config done.")
+        self.unit.status = MaintenanceStatus("Removing supplementary "
+                                             "Kubernetes objects")
+        utils.remove_k8s_objects(self._stored.namespace)
+        self.unit.status = MaintenanceStatus("Removing pod")
```
Comment (camille-rodriguez) [inline charms/metallb-speaker/src/charm.py:86]: I would also set self._stored.set_default(k8s_objects_created=False) here

File: charms/metallb-controller/src/charm.py [hunk 2]
```diff
@@ -21,93 +21,84 @@
...
-        utils.delete_namespaced_role_with_api(
-            name='config-watcher',
-            namespace=self._stored.namespace
-        )
-        self.model.unit.status = ActiveStatus("Removing extra config done.")
-        self._stored.configured = False
+        self.unit.status = MaintenanceStatus("Removing supplementary "
+                                             "Kubernetes objects")
+        utils.remove_k8s_objects(self._stored.namespace)
+        self.unit.status = MaintenanceStatus("Removing pod")
```
Comment (camille-rodriguez) [inline charms/metallb-controller/src/charm.py:92]: I would also set self._stored.set_default(k8s_objects_created=False) when the remove hook is triggered

File: charms/metallb-speaker/src/charm.py [hunk 3]
```diff
@@ -22,97 +22,72 @@
...
         # -- initialize states --
+        self._stored.set_default(k8s_objects_created=False)
         self._stored.set_default(started=False)
+        self._stored.set_default(
+            secret=b64encode(
+                utils._random_secret(128).encode('utf-8')
+            ).decode('utf-8'))
         # -- base values --
         self._stored.set_default(namespace=os.environ["JUJU_MODEL_NAME"])
         self._stored.set_default(container_image='metallb/speaker:v0.9.3')
```
Comment (camille-rodriguez) [inline charms/metallb-speaker/src/charm.py:50]: I do not think that this is used anymore.

**Review by camille-rodriguez — COMMENTED** [review]
Good job. A few minor suggestions, otherwise it looks great!

File: charms/metallb-controller/src/charm.py [hunk 4]
```diff
@@ -21,93 +21,84 @@
...
         """Charm initialization for events observation."""
         super().__init__(*args)
-        if not self.model.unit.is_leader():
-            self.model.unit.status = WaitingStatus("Waiting for leadership")
+        if not self.unit.is_leader():
+            self.unit.status = WaitingStatus("Waiting for leadership")
             return
         self.image = OCIImageResource(self, 'metallb-controller-image')
-        self.framework.observe(self.on.start, self.on_start)
+        self.framework.observe(self.on.install, self._on_start)
```
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-controller/src/charm.py:36]: Ah, we definitely should include `leader-elected`.  In theory, the first unit should have leadership from the get-go because there's no other unit to have claimed it already.  I would consider it not having everything it needs during the install hook to be a bug, just like [sometimes not having resources](https://bugs.launchpad.net/juju/+bug/1763175) or [the install hook not being called at all](https://discourse.juju.is/t/install-hook-now-called-for-kubernetes-charms/2585/2).  So, while technically it most likely won't do anything during the install hook, I was including that because in an ideal world that would lead to the fastest startup time for the workload.

File: charms/metallb-controller/src/charm.py [hunk 5]
```diff
@@ -21,93 +21,84 @@
...
+        self.unit.status = MaintenanceStatus("Removing supplementary "
+                                             "Kubernetes objects")
+        utils.remove_k8s_objects(self._stored.namespace)
+        self.unit.status = MaintenanceStatus("Removing pod")
         self._stored.started = False
 
-    def set_pod_spec(self):
+    def set_pod_spec(self, image_info):
         """Set pod spec."""
         iprange = self.model.config["iprange"].split(",")
```
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-controller/src/charm.py:97]: NB: storing the value was just done to avoid unnecessary restarts of the workload.  If the value were larger, or if there were multiple values which might lead to a change, I'd probably just store a hash instead.  I might go ahead and do that anyway, just for future-proofing.

File: charms/metallb-controller/tests/requirements.txt [hunk 6]
```diff
@@ -1,3 +0,0 @@
-mock
```
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-controller/tests/requirements.txt:1]: They're managed by pipenv now, and included in the `Pipfile`.  The only reason I'm keeping the `requirements.txt` file around is for `charmbuild`, but I plan to propose a PR to support the `Pipfile.lock`

File: charms/metallb-speaker/Pipfile.lock [hunk 7]
```diff
@@ -0,0 +1,526 @@
+{
```
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-speaker/Pipfile.lock:1]: It comes from [pipenv](https://pipenv.pypa.io/en/latest/) which is the [recommended dependency manager](https://packaging.python.org/guides/tool-recommendations/) now.  It's [generally recommended](https://pipenv.pypa.io/en/latest/basics/#general-recommendations-version-control) to keep the `Pipfile.lock` in version control, unless you're targeting multiple versions of Python; with `charmcraft`, I think the Python version should be locked at build time (via the venv), but that may need to be tested.  For K8s charms, however, the Python version is set by the Juju unit operator image, anyway.

File: charms/metallb-speaker/requirements.txt [hunk 8]
```diff
@@ -1,3 +1,20 @@
-ops
-kubernetes
-https://github.com/juju-solutions/resource-oci-image/archive/master.zip
\ No newline at end of file
+-i https://pypi.org/simple
+cachetools==4.1.1; python_version ~= '3.5'
+certifi==2020.6.20
+chardet==3.0.4
+git+https://github.com/canonical/operator/@e3309464a0cbd8f9467199243721c9fb9d91271d#egg=ops
```
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-speaker/requirements.txt:5]: This is locked to a specific commit and requires manual intervention (calling `make update`) to move it forward.  As discussed in chat yesterday, this was to pull in a specific fix that hadn't been released yet, but now that 0.10.0 is out, I'll switch this back to PyPI.

File: charms/metallb-speaker/src/charm.py (hunk 1 above)
Comment (johnsca) [reply charms/metallb-speaker/src/charm.py:86]: Excellent catch.

File: charms/metallb-controller/src/charm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-controller/src/charm.py:36]: Oh, I should also note that if the charm doesn't have leadership, then the observe won't even be called, so it will be exactly the same as not having it there (no extra work done) but including it allows us the possibility to get the workload up sooner if there's some chance that we can.

File: charms/metallb-speaker/Pipfile.lock (hunk 7 above)
[the remark this answers is not in the document]
Comment (johnsca) [reply charms/metallb-speaker/Pipfile.lock:1]: Hrm.  It seems that I was mistaken and [`charmcraft` doesn't bundle or lock the Python version](https://discourse.juju.is/t/what-files-would-you-expect-charmcraft-build-copies-into-the-charm/3247/12).  I still think that including the `Pipfile.lock` in version control makes more sense for our use-case, though, since I think the consensus is that updating dependencies should be an explicit step rather than something that happens at every build.  (The reactive charm build process follows the latter approach and it has required charm authors to jump through quite a few hoops to work around that to ensure the ability to rebuild a previous release to add patches, for example.)

File: charms/metallb-controller/src/charm.py [hunk 9]
```diff
@@ -21,93 +23,93 @@
...
         self._stored.set_default(started=False)
-        self._stored.set_default(configured=False)
+        self._stored.set_default(config_hash=self._config_hash())
         # -- base values --
         self._stored.set_default(namespace=os.environ["JUJU_MODEL_NAME"])
 
-    def _on_config_changed(self, event):
-        if not self._stored.started:
+    def _config_hash(self):
+        data = json.dumps({
```
Comment (camille-rodriguez) [inline charms/metallb-controller/src/charm.py:52]: Why is this needed instead of grabbing directly the model config ?

Comment (camille-rodriguez) [reply charms/metallb-controller/src/charm.py:52]: Ah, I imagine to compare all the config items instead of comparing them one by one to the stored object

**Comment by camille-rodriguez** [comment]
Looks good to me!

File: charms/metallb-controller/src/charm.py (hunk 9 above)
Comment (johnsca) [reply charms/metallb-controller/src/charm.py:52]: Yeah, this is more future-proofing for when we want to handle changes to the `protocol` config as well.  I also just generally consider it cleaner to not store a full copy of the data just to tell if it changed.
